File: mv-design-pro/frontend/src/ui/sld/core/geometryOverrides.ts
/**
 * ProjectGeometryOverridesV1 — kontrakt nadpisan geometrii SLD (zamrozony v1).
 *
 * CANONICAL CONTRACT (BINDING — RUN #3H):
 * - Overrides sa OSOBNA warstwa — NIE sa czescia LayoutEngine.
 * - LayoutResultV1 pozostaje niezmieniony (hash-stabilny).
 * - Skladanie: applyOverrides(LayoutResultV1, OverridesV1) → EffectiveLayoutV1.
 * - Hashing: canonical SHA-256 (permutation invariant, bez timestampow).
 * - Walidacja: elementId musi istniec w LayoutResult; brak kolizji; port constraints.
 *
 * PIPELINE:
 *   LayoutResultV1 + ProjectGeometryOverridesV1 → applyOverrides → EffectiveLayoutV1
 */

// =============================================================================
// VERSION
// =============================================================================

export const OVERRIDES_VERSION = '1.0' as const;

// =============================================================================
// SCOPE — zakres nadpisania
// =============================================================================

export const OverrideScopeV1 = {
  /** Wezel (source, load, junction) */
  NODE: 'NODE',
  /** Blok stacji (SwitchgearBlockV1) */
  BLOCK: 'BLOCK',
  /** Pole w bloku stacji */
  FIELD: 'FIELD',
  /** Etykieta elementu */
  LABEL: 'LABEL',
  /** Kanal krawedzi (reserved — nie zaimplementowany w v1) */
  EDGE_CHANNEL: 'EDGE_CHANNEL',
} as const;

export type OverrideScopeV1 = (typeof OverrideScopeV1)[keyof typeof OverrideScopeV1];

// =============================================================================
// OPERATION — typ operacji geometrycznej
// =============================================================================

export const OverrideOperationV1 = {
  /** Przesuniecie wzgledem pozycji bazowej (delta x,y) */
  MOVE_DELTA: 'MOVE_DELTA',
  /** Wymuszenie kolejnosci pol w bloku stacji */
  REORDER_FIELD: 'REORDER_FIELD',
  /** Wymuszenie pozycji etykiety */
  MOVE_LABEL: 'MOVE_LABEL',
} as const;

export type OverrideOperationV1 = (typeof OverrideOperationV1)[keyof typeof OverrideOperationV1];

// =============================================================================
// PAYLOAD — dane per operacja (walidowane)
// =============================================================================

/** Payload dla MOVE_DELTA: przesuniecie wzgledem bazy. */
export interface MoveDeltaPayloadV1 {
  readonly dx: number;
  readonly dy: number;
}

/** Payload dla REORDER_FIELD: kolejnosc pol. */
export interface ReorderFieldPayloadV1 {
  readonly fieldOrder: readonly string[];
}

/** Payload dla MOVE_LABEL: pozycja etykiety. */
export interface MoveLabelPayloadV1 {
  readonly anchorX: number;
  readonly anchorY: number;
}

/** Unia payloadow — walidowana per operacja. */
export type GeometryOverridePayloadV1 =
  | MoveDeltaPayloadV1
  | ReorderFieldPayloadV1
  | MoveLabelPayloadV1;

// =============================================================================
// OVERRIDE ITEM — pojedynczy rekord nadpisania
// =============================================================================

export interface GeometryOverrideItemV1 {
  /** Globalny identyfikator elementu (= nodeId / blockId / fieldId). */
  readonly elementId: string;
  /** Zakres nadpisania. */
  readonly scope: OverrideScopeV1;
  /** Typ operacji. */
  readonly operation: OverrideOperationV1;
  /** Dane operacji (walidowane per operation). */
  readonly payload: GeometryOverridePayloadV1;
}

// =============================================================================
// TOP-LEVEL CONTRACT
// =============================================================================

/**
 * ProjectGeometryOverridesV1 — zamrozony kontrakt nadpisan geometrii.
 *
 * INVARIANTY:
 * 1. items sortowane deterministycznie po (elementId, scope, operation).
 * 2. Hash liczony WYLACZNIE z items (bez timestampow, bez identyfikatorow sesji).
 * 3. Permutation invariance — kolejnosc dostarczenia nie ma znaczenia.
 * 4. overridesVersion = '1.0' (zamrozony).
 */
export interface ProjectGeometryOverridesV1 {
  /** Wersja kontraktu nadpisan. */
  readonly overridesVersion: typeof OVERRIDES_VERSION;
  /** ID studium przypadku. */
  readonly studyCaseId: string;
  /** Hash snapshotu ENM (referencja do stanu modelu). */
  readonly snapshotHash: string;
  /** Lista nadpisan — sortowana deterministycznie. */
  readonly items: readonly GeometryOverrideItemV1[];
}

// =============================================================================
// EMPTY OVERRIDES
// =============================================================================

/** Puste overrides (brak nadpisan). */
export function emptyOverrides(studyCaseId: string, snapshotHash: string): ProjectGeometryOverridesV1 {
  return {
    overridesVersion: OVERRIDES_VERSION,
    studyCaseId,
    snapshotHash,
    items: [],
  };
}

// =============================================================================
// CANONICAL SERIALIZATION
// =============================================================================

/**
 * Kanonizuje overrides — sortuje items deterministycznie.
 *
 * Sortowanie: elementId → scope → operation (leksykograficznie).
 */
export function canonicalizeOverrides(
  overrides: ProjectGeometryOverridesV1,
): ProjectGeometryOverridesV1 {
  const sorted = [...overrides.items].sort((a, b) => {
    const cmp1 = a.elementId.localeCompare(b.elementId);
    if (cmp1 !== 0) return cmp1;
    const cmp2 = a.scope.localeCompare(b.scope);
    if (cmp2 !== 0) return cmp2;
    return a.operation.localeCompare(b.operation);
  });

  return {
    overridesVersion: overrides.overridesVersion,
    studyCaseId: overrides.studyCaseId,
    snapshotHash: overrides.snapshotHash,
    items: sorted,
  };
}

/**
 * Oblicza deterministyczny hash nadpisan.
 *
 * Hash liczy sie WYLACZNIE z items (kanoniczny JSON):
 * - Klucze JSON w kolejnosci alfabetycznej.
 * - BEZ timestampow, identyfikatorow sesji, createdAt.
 * - Permutation invariant (sort po elementId/scope/operation).
 *
 * Algorytm: FNV-1a 32-bit (spojny z computeLayoutResultHash).
 */
export function computeOverridesHash(overrides: ProjectGeometryOverridesV1): string {
  const canonical = canonicalizeOverrides(overrides);

  const hashInput = canonical.items.map((item) => ({
    elementId: item.elementId,
    operation: item.operation,
    payload: item.payload,
    scope: item.scope,
  }));

  const json = JSON.stringify(hashInput);

  // FNV-1a 32-bit hash (ten sam algorytm co computeLayoutResultHash)
  let hash = 0x811c9dc5;
  for (let i = 0; i < json.length; i++) {
    hash ^= json.charCodeAt(i);
    hash = Math.imul(hash, 0x01000193);
  }

  return (hash >>> 0).toString(16).padStart(8, '0');
}

// =============================================================================
// GRID SNAP
// =============================================================================

/** Domyslny rozmiar siatki snap-to-grid (w px). */
export const GEOMETRY_GRID_SNAP = 20;

/**
 * Przyciaga wartosc do siatki (deterministyczne).
 */
export function snapToGrid(value: number, gridSize: number = GEOMETRY_GRID_SNAP): number {
  return Math.round(value / gridSize) * gridSize;
}

/**
 * Przyciaga payload MOVE_DELTA do siatki.
 */
export function snapDeltaToGrid(
  dx: number,
  dy: number,
  gridSize: number = GEOMETRY_GRID_SNAP,
): MoveDeltaPayloadV1 {
  return {
    dx: snapToGrid(dx, gridSize),
    dy: snapToGrid(dy, gridSize),
  };
}

// =============================================================================
// VALIDATION TYPES
// =============================================================================

export interface OverrideValidationErrorV1 {
  readonly elementId: string;
  readonly code: string;
  readonly message: string;
}

export interface OverrideValidationResultV1 {
  readonly valid: boolean;
  readonly errors: readonly OverrideValidationErrorV1[];
}

// =============================================================================
// VALIDATION
// =============================================================================

/**
 * Waliduje overrides przeciwko LayoutResultV1.
 *
 * Reguly:
 * 1. elementId musi istniec w LayoutResult (jako nodeId lub blockId).
 * 2. MOVE_DELTA: payload musi miec dx/dy jako liczby.
 * 3. REORDER_FIELD: fieldOrder musi byc tablica stringow.
 * 4. MOVE_LABEL: anchorX/anchorY musi byc liczbami.
 * 5. Scope musi byc poprawny dla danej operacji.
 *
 * Kolizje (overlap) i port constraints sprawdzane sa w applyOverrides()
 * (po nalozeniu — bo wymagaja pelnego kontekstu geometrii).
 */
export function validateOverridesAgainstLayout(
  overrides: ProjectGeometryOverridesV1,
  layoutNodeIds: ReadonlySet<string>,
  layoutBlockIds: ReadonlySet<string>,
): OverrideValidationResultV1 {
  const errors: OverrideValidationErrorV1[] = [];

  for (const item of overrides.items) {
    // Rule 1: elementId musi istniec
    const existsInNodes = layoutNodeIds.has(item.elementId);
    const existsInBlocks = layoutBlockIds.has(item.elementId);

    if (item.scope === OverrideScopeV1.NODE && !existsInNodes) {
      errors.push({
        elementId: item.elementId,
        code: 'geometry.override_invalid_element',
        message: `Element '${item.elementId}' nie istnieje w layoucie (NODE scope)`,
      });
      continue;
    }

    if (item.scope === OverrideScopeV1.BLOCK && !existsInBlocks) {
      errors.push({
        elementId: item.elementId,
        code: 'geometry.override_invalid_element',
        message: `Blok '${item.elementId}' nie istnieje w layoucie (BLOCK scope)`,
      });
      continue;
    }

    if (item.scope === OverrideScopeV1.LABEL && !existsInNodes && !existsInBlocks) {
      errors.push({
        elementId: item.elementId,
        code: 'geometry.override_invalid_element',
        message: `Element '${item.elementId}' nie istnieje w layoucie (LABEL scope)`,
      });
      continue;
    }

    // Rule 2: MOVE_DELTA payload
    if (item.operation === OverrideOperationV1.MOVE_DELTA) {
      const p = item.payload as MoveDeltaPayloadV1;
      if (typeof p.dx !== 'number' || typeof p.dy !== 'number') {
        errors.push({
          elementId: item.elementId,
          code: 'geometry.override_invalid_element',
          message: `MOVE_DELTA: payload musi miec dx/dy jako liczby`,
        });
      }
      if (!isFinite(p.dx) || !isFinite(p.dy)) {
        errors.push({
          elementId: item.elementId,
          code: 'geometry.override_invalid_element',
          message: `MOVE_DELTA: dx/dy musza byc skonczne`,
        });
      }
    }

    // Rule 3: REORDER_FIELD payload
    if (item.operation === OverrideOperationV1.REORDER_FIELD) {
      const p = item.payload as ReorderFieldPayloadV1;
      if (!Array.isArray(p.fieldOrder)) {
        errors.push({
          elementId: item.elementId,
          code: 'geometry.override_forbidden_for_station_type',
          message: `REORDER_FIELD: fieldOrder musi byc tablica`,
        });
      }
    }

    // Rule 4: MOVE_LABEL payload
    if (item.operation === OverrideOperationV1.MOVE_LABEL) {
      const p = item.payload as MoveLabelPayloadV1;
      if (typeof p.anchorX !== 'number' || typeof p.anchorY !== 'number') {
        errors.push({
          elementId: item.elementId,
          code: 'geometry.override_invalid_element',
          message: `MOVE_LABEL: anchorX/anchorY musza byc liczbami`,
        });
      }
    }

    // Rule 5: scope-operation compatibility
    if (
      item.scope === OverrideScopeV1.NODE &&
      item.operation !== OverrideOperationV1.MOVE_DELTA
    ) {
      errors.push({
        elementId: item.elementId,
        code: 'geometry.override_forbidden_for_station_type',
        message: `NODE scope dozwala tylko operacji MOVE_DELTA`,
      });
    }

    if (
      item.scope === OverrideScopeV1.FIELD &&
      item.operation !== OverrideOperationV1.REORDER_FIELD
    ) {
      errors.push({
        elementId: item.elementId,
        code: 'geometry.override_forbidden_for_station_type',
        message: `FIELD scope dozwala tylko operacji REORDER_FIELD`,
      });
    }

    if (
      item.scope === OverrideScopeV1.LABEL &&
      item.operation !== OverrideOperationV1.MOVE_LABEL
    ) {
      errors.push({
        elementId: item.elementId,
        code: 'geometry.override_forbidden_for_station_type',
        message: `LABEL scope dozwala tylko operacji MOVE_LABEL`,
      });
    }
  }

  return { valid: errors.length === 0, errors };
}

// =============================================================================
// FIXACTION CODES (CAD)
// =============================================================================

/**
 * Kody FixActions dla trybu projektowego (CAD).
 * Stabilne identyfikatory — nigdy nie zmieniac istniejacych.
 */
export const GeometryFixCodes = {
  OVERRIDE_INVALID_ELEMENT: 'geometry.override_invalid_element',
  OVERRIDE_CAUSES_COLLISION: 'geometry.override_causes_collision',
  OVERRIDE_BREAKS_PORT_CONSTRAINTS: 'geometry.override_breaks_port_constraints',
  OVERRIDE_FORBIDDEN_FOR_STATION_TYPE: 'geometry.override_forbidden_for_station_type',
  OVERRIDE_REQUIRES_UNLOCK: 'geometry.override_requires_unlock',
} as const;

export type GeometryFixCode = (typeof GeometryFixCodes)[keyof typeof GeometryFixCodes];

File: mv-design-pro/backend/src/domain/geometry_overrides.py
"""
ProjectGeometryOverridesV1 — domain model for SLD geometry overrides.

CANONICAL CONTRACT (BINDING — RUN #3H):
- Overrides are a SEPARATE layer — NOT part of LayoutEngine.
- LayoutResultV1 remains unchanged (hash-stable).
- Hashing: canonical SHA-256 (permutation invariant, no timestamps).
- Validation: elementId must exist in LayoutResult; no collisions.

PIPELINE:
  LayoutResultV1 + ProjectGeometryOverridesV1 → applyOverrides → EffectiveLayoutV1
"""

from __future__ import annotations

import hashlib
import json
from dataclasses import dataclass, field
from enum import Enum
from typing import Any


# =============================================================================
# VERSION
# =============================================================================

OVERRIDES_VERSION = "1.0"


# =============================================================================
# ENUMS
# =============================================================================


class OverrideScopeV1(str, Enum):
    """Scope (zakres) nadpisania geometrii."""

    NODE = "NODE"
    BLOCK = "BLOCK"
    FIELD = "FIELD"
    LABEL = "LABEL"
    EDGE_CHANNEL = "EDGE_CHANNEL"


class OverrideOperationV1(str, Enum):
    """Typ operacji geometrycznej."""

    MOVE_DELTA = "MOVE_DELTA"
    REORDER_FIELD = "REORDER_FIELD"
    MOVE_LABEL = "MOVE_LABEL"


# =============================================================================
# OVERRIDE ITEM
# =============================================================================


@dataclass(frozen=True)
class GeometryOverrideItemV1:
    """Pojedynczy rekord nadpisania geometrii."""

    element_id: str
    scope: OverrideScopeV1
    operation: OverrideOperationV1
    payload: dict[str, Any]

    def to_dict(self) -> dict[str, Any]:
        """Serializacja do dict (kanoniczny format)."""
        return {
            "element_id": self.element_id,
            "operation": self.operation.value,
            "payload": self.payload,
            "scope": self.scope.value,
        }


# =============================================================================
# TOP-LEVEL CONTRACT
# =============================================================================


@dataclass(frozen=True)
class ProjectGeometryOverridesV1:
    """Zamrozony kontrakt nadpisan geometrii SLD."""

    overrides_version: str = OVERRIDES_VERSION
    study_case_id: str = ""
    snapshot_hash: str = ""
    items: tuple[GeometryOverrideItemV1, ...] = field(default_factory=tuple)

    def to_dict(self) -> dict[str, Any]:
        """Serializacja do dict."""
        return {
            "overrides_version": self.overrides_version,
            "study_case_id": self.study_case_id,
            "snapshot_hash": self.snapshot_hash,
            "items": [item.to_dict() for item in self.items],
        }

    @staticmethod
    def from_dict(data: dict[str, Any]) -> ProjectGeometryOverridesV1:
        """Deserializacja z dict."""
        items = tuple(
            GeometryOverrideItemV1(
                element_id=item["element_id"],
                scope=OverrideScopeV1(item["scope"]),
                operation=OverrideOperationV1(item["operation"]),
                payload=item.get("payload", {}),
            )
            for item in data.get("items", [])
        )
        return ProjectGeometryOverridesV1(
            overrides_version=data.get("overrides_version", OVERRIDES_VERSION),
            study_case_id=data.get("study_case_id", ""),
            snapshot_hash=data.get("snapshot_hash", ""),
            items=items,
        )


# =============================================================================
# CANONICAL SERIALIZATION + HASHING
# =============================================================================


def canonicalize_overrides(
    overrides: ProjectGeometryOverridesV1,
) -> ProjectGeometryOverridesV1:
    """Kanonizuje overrides — sortuje items deterministycznie.

    Sortowanie: element_id → scope → operation (leksykograficznie).
    """
    sorted_items = tuple(
        sorted(
            overrides.items,
            key=lambda item: (
                item.element_id,
                item.scope.value,
                item.operation.value,
            ),
        )
    )
    return ProjectGeometryOverridesV1(
        overrides_version=overrides.overrides_version,
        study_case_id=overrides.study_case_id,
        snapshot_hash=overrides.snapshot_hash,
        items=sorted_items,
    )


def compute_overrides_hash(overrides: ProjectGeometryOverridesV1) -> str:
    """Oblicza deterministyczny SHA-256 hash nadpisan.

    Hash liczy sie WYLACZNIE z items (kanoniczny JSON):
    - Klucze JSON w kolejnosci alfabetycznej (sort_keys=True).
    - BEZ timestampow, identyfikatorow sesji.
    - Permutation invariant (sort po element_id/scope/operation).
    """
    canonical = canonicalize_overrides(overrides)
    items_data = [item.to_dict() for item in canonical.items]
    json_str = json.dumps(items_data, sort_keys=True, separators=(",", ":"))
    return hashlib.sha256(json_str.encode("utf-8")).hexdigest()


# =============================================================================
# VALIDATION
# =============================================================================


@dataclass(frozen=True)
class OverrideValidationErrorV1:
    """Blad walidacji nadpisania."""

    element_id: str
    code: str
    message: str


@dataclass(frozen=True)
class OverrideValidationResultV1:
    """Wynik walidacji nadpisan."""

    valid: bool
    errors: tuple[OverrideValidationErrorV1, ...] = field(default_factory=tuple)


SCOPE_ALLOWED_OPERATIONS: dict[OverrideScopeV1, set[OverrideOperationV1]] = {
    OverrideScopeV1.NODE: {OverrideOperationV1.MOVE_DELTA},
    OverrideScopeV1.BLOCK: {OverrideOperationV1.MOVE_DELTA},
    OverrideScopeV1.FIELD: {OverrideOperationV1.REORDER_FIELD},
    OverrideScopeV1.LABEL: {OverrideOperationV1.MOVE_LABEL},
    OverrideScopeV1.EDGE_CHANNEL: set(),  # reserved
}


def validate_overrides(
    overrides: ProjectGeometryOverridesV1,
    known_node_ids: set[str] | frozenset[str],
    known_block_ids: set[str] | frozenset[str],
) -> OverrideValidationResultV1:
    """Waliduje overrides przeciwko znanym elementom layoutu.

    Reguly:
    1. element_id musi istniec w layoucie (node lub block).
    2. Operacja musi byc dozwolona dla danego scope.
    3. Payload musi miec wymagane pola.
    """
    errors: list[OverrideValidationErrorV1] = []

    for item in overrides.items:
        # Rule 1: element_id musi istniec
        if item.scope == OverrideScopeV1.NODE:
            if item.element_id not in known_node_ids:
                errors.append(
                    OverrideValidationErrorV1(
                        element_id=item.element_id,
                        code="geometry.override_invalid_element",
                        message=f"Element '{item.element_id}' nie istnieje (NODE scope)",
                    )
                )
                continue
        elif item.scope == OverrideScopeV1.BLOCK:
            if item.element_id not in known_block_ids:
                errors.append(
                    OverrideValidationErrorV1(
                        element_id=item.element_id,
                        code="geometry.override_invalid_element",
                        message=f"Blok '{item.element_id}' nie istnieje (BLOCK scope)",
                    )
                )
                continue
        elif item.scope == OverrideScopeV1.LABEL:
            if (
                item.element_id not in known_node_ids
                and item.element_id not in known_block_ids
            ):
                errors.append(
                    OverrideValidationErrorV1(
                        element_id=item.element_id,
                        code="geometry.override_invalid_element",
                        message=f"Element '{item.element_id}' nie istnieje (LABEL scope)",
                    )
                )
                continue

        # Rule 2: scope-operation compatibility
        allowed = SCOPE_ALLOWED_OPERATIONS.get(item.scope, set())
        if item.operation not in allowed:
            errors.append(
                OverrideValidationErrorV1(
                    element_id=item.element_id,
                    code="geometry.override_forbidden_for_station_type",
                    message=(
                        f"Operacja {item.operation.value} niedozwolona "
                        f"dla scope {item.scope.value}"
                    ),
                )
            )
            continue

        # Rule 3: payload validation
        if item.operation == OverrideOperationV1.MOVE_DELTA:
            if "dx" not in item.payload or "dy" not in item.payload:
                errors.append(
                    OverrideValidationErrorV1(
                        element_id=item.element_id,
                        code="geometry.override_invalid_element",
                        message="MOVE_DELTA: payload musi miec dx/dy",
                    )
                )
        elif item.operation == OverrideOperationV1.MOVE_LABEL:
            if "anchorX" not in item.payload or "anchorY" not in item.payload:
                errors.append(
                    OverrideValidationErrorV1(
                        element_id=item.element_id,
                        code="geometry.override_invalid_element",
                        message="MOVE_LABEL: payload musi miec anchorX/anchorY",
                    )
                )
        elif item.operation == OverrideOperationV1.REORDER_FIELD:
            if "fieldOrder" not in item.payload:
                errors.append(
                    OverrideValidationErrorV1(
                        element_id=item.element_id,
                        code="geometry.override_forbidden_for_station_type",
                        message="REORDER_FIELD: payload musi miec fieldOrder",
                    )
                )

    return OverrideValidationResultV1(
        valid=len(errors) == 0,
        errors=tuple(errors),
    )


# =============================================================================
# FIXACTION CODES (CAD)
# =============================================================================


class GeometryFixCode(str, Enum):
    """Stabilne kody FixActions dla trybu projektowego (CAD)."""

    OVERRIDE_INVALID_ELEMENT = "geometry.override_invalid_element"
    OVERRIDE_CAUSES_COLLISION = "geometry.override_causes_collision"
    OVERRIDE_BREAKS_PORT_CONSTRAINTS = "geometry.override_breaks_port_constraints"
    OVERRIDE_FORBIDDEN_FOR_STATION_TYPE = "geometry.override_forbidden_for_station_type"
    OVERRIDE_REQUIRES_UNLOCK = "geometry.override_requires_unlock"
